### ISSUE TITLE
Write to removable (external) sd card fails

### ISSUE THREAD

**Issue by metalurgus** [spacecowboy/NoNonsense-FilePicker#98]
Starting from Android 4.4 Google restricted write to removable storage, and introduced Storage Access Framwork (SAF), which is the only way I know to write on external sd card.    
The problem is that the library does not use SAF, and though does not have permissions to write to removable storage

**Comment by spacecowboy** [comment]
The library can't define any permissions. You have to do that yourself in your app which includes it. SAF is not required.

An example is the sample app, which defines that it requires the `WRITE_EXTERNAL_STORAGE` permission:
https://github.com/spacecowboy/NoNonsense-FilePicker/blob/master/sample/src/main/AndroidManifest.xml#L12

By default, the sample app will start you in the internal memory. You will have to navigate to the directory for the external SD card. In your own app you can start the picker in any directory you want of course.

Android docs on this are located here: https://developer.android.com/guide/topics/data/data-storage.html#filesExternal

**Comment by metalurgus** [comment]
Ok, I'll try to explain the problem once more: 
Some devices have a slot for microSD memory card. If you insert the card, it will be mounted as external memory card, usually as `/storage/sdcard1`, while embedded emulated sd card is mounted as  `/storage/sdcard0`, and is used as `EXTERNAL_STORAGE`. 
But if I try to create directory inside `/storage/sdcard1`, it fails showing toast : `Failed to create folder`. [Link to source](https://github.com/spacecowboy/NoNonsense-FilePicker/blob/master/library/src/main/java/com/nononsenseapps/filepicker/FilePickerFragment.java#L309). This happens because beginning from some Android version, google introduced official support of external SD card, and the only way to write there that I know is SAF.

**Comment by spacecowboy** [comment]
@metalurgus I understood the issue the first time, don't worry :)

The docs claim that there shouldn't be any difference, but if it doesn't work for you, then obviously there is something the matter.

Anyway, thanks for the clarification, I'll try and investigate.

**Comment by metalurgus** [comment]
I actually can't find in documentation, but everyone over the internet is talking that in Android 4.4 google totally restricted access to external SD card. But in 5.0 they introduced SAF, which allows write access to external SD card. I've tested on Sony Xperia m2 (D2305) Android 5.1.1, and Samsung Galaxy J5 having a slot for SD card, and both can't create a directory inside external SD card, Also I have about 50 reports from users that they face the same problem when they try to create directory in the external SD card using my app. 
So it is definately not a single device problem. 

Also while investigating the problem, I tried ES File Explorer,  and I see that it is using SAF to write to external SD card

**Comment by spacecowboy** [comment]
Using SAF in the library is out of the question (and kind of defeats the point of the library.) Just imagine using one picker first, in order to open a second. A bit redundant ;) 

Perhaps the correct thing to do is to use SAF in your app if running on version >= 5.0?

**Comment by metalurgus** [comment]
You are right, but how about the functionality to create a directory of the library?

**Comment by spacecowboy** [comment]
Ah yes, that's right. That was one of my reasons for making the library in the first place.

I'll try and see if I can find a solution.

**Comment by mianharisali** [comment]
facing same issue :(

**Comment by sregg** [comment]
Same here.
Any update on this issue?

**Comment by metalurgus** [comment]
@sregg yes I did. Here is my [question ](http://stackoverflow.com/questions/40172011/storage-access-framework-takepersistableuripermission)on the StackOverflow, and [my own answer](http://stackoverflow.com/a/40204553/1568530) for this question

**Comment by AdityaAnand1** [comment]
Oh my god, I can't believe how long it took me to pin this down. This issue right here is why I got 100+ 1 stars on my app.

Has there been any progress? @spacecowboy? @metalurgus What kinda solution did you build?

Anyways, the readme should at least mention in bold letters that you cannot create folders on external SD card with this library. 

Thanks Google.

**Comment by spacecowboy** [comment]
See https://gitlab.com/spacecowboy/Feeder/blob/master/app/src/main/kotlin/com/nononsenseapps/feeder/ui/FeedActivity.kt for my preferred solution. (Use Android's built-in picker for Kitkat+ devices and only use this library as a fallback for pre-kitkat devices).

You'll technically have two pickers in your app but users will only see one of them. And Kitkat+ will be consistent with other apps.
